src/jrd/replication/Publisher.cpp

```diff
@@ -226,74 +244,100 @@ namespace
 	}
 
 	class ReplicatedRecordImpl :
-		public AutoIface<IReplicatedRecordImpl<ReplicatedRecordImpl, CheckStatusWrapper> >
+		public AutoIface<IReplicatedRecordImpl<ReplicatedRecordImpl, CheckStatusWrapper> >,
+		public IReplicatedFieldImpl<ReplicatedRecordImpl, CheckStatusWrapper>
 	{
 	public:
-		ReplicatedRecordImpl(thread_db* tdbb, const Record* record)
+		ReplicatedRecordImpl(thread_db* tdbb, const jrd_rel* relation, const Record* record)
 			: //m_tdbb(tdbb),
-			  m_record(record)
+			  m_record(record),
+			  m_format(record->getFormat()),
+			  m_relation(relation)
 		{
 		}
 
 		~ReplicatedRecordImpl()
 		{
 		}
 
-		unsigned getRawLength()
+		unsigned getCount() override
 		{
-			return m_record->getLength();
+			return m_format->fmt_count;
 		}
 
-		const unsigned char* getRawData()
+		const char* getName() override
 		{
-			return m_record->getData();
+			jrd_fld* field = MET_get_field(m_relation, m_fieldIndex);
+
+			if (field == nullptr)
+				return nullptr;
+
+			return field->fld_name.c_str();
 		}
 
-	private:
-		//thread_db* const m_tdbb;
-		const Record* const m_record;
-	};
+		unsigned getType() override
+		{
+			return m_fieldType;
+		}
 
-	class ReplicatedBlobImpl :
-		public AutoIface<IReplicatedBlobImpl<ReplicatedBlobImpl, CheckStatusWrapper> >
-	{
-	public:
-		ReplicatedBlobImpl(thread_db* tdbb, jrd_tra* transaction, const bid* blobId) :
-			m_tdbb(tdbb), m_blob(blb::open(tdbb, transaction, blobId))
+		unsigned getSubType() override
 		{
+			return m_format->fmt_desc[m_fieldIndex].getSubType();
 		}
 
-		~ReplicatedBlobImpl()
+		unsigned getScale() override
 		{
-			m_blob->BLB_close(m_tdbb);
+			return m_format->fmt_desc[m_fieldIndex].dsc_scale;
 		}
 
-		unsigned getLength()
+		unsigned getLength() override
 		{
-			return m_blob->blb_length;
+			return m_fieldLength;
 		}
 
-		FB_BOOLEAN isEof()
+		unsigned getCharSet() override
 		{
-			return (m_blob->blb_flags & BLB_eof);
+			return m_format->fmt_desc[m_fieldIndex].getCharSet();
 		}
 
-		unsigned getSegment(unsigned length, unsigned char* buffer)
+		const void* getData() override
 		{
-			auto p = buffer;
+			if (m_record->isNull(m_fieldIndex))
+				return nullptr;
 
-			if (length && !(m_blob->blb_flags & BLB_eof))
-			{
-				const auto n = (USHORT) MIN(length, MAX_USHORT);
-				p += m_blob->BLB_get_segment(m_tdbb, p, n);
-			}
+			return m_record->getData() + (IPTR)(m_format->fmt_desc[m_fieldIndex].dsc_address);
+		}
+
+		IReplicatedField* getField(unsigned index) override
+		{
+			if (index >= m_format->fmt_count)
+				return nullptr;
+
+			m_fieldIndex = index;
+			SLONG dummySubtype, dummyScale;
+			m_format->fmt_desc[m_fieldIndex].getSqlInfo(&m_fieldLength, &dummySubtype, &dummyScale, &m_fieldType);
```

Comment:
Why not including subtype and scale to the record object too? Stack allocation is dirty cheap and returning the value should be slightly faster than m_format->fmt_desc[m_fieldIndex] redirection + method call.

Reply:
They are involved in charset trick. Calling existing routines is better than copy-pasting them in this case.